### ISSUE TITLE
Hrnet with LOVASZ_SOFTMAX_LOSS error

### ISSUE THREAD

**Issue by lytk01** [PaddlePaddle/PaddleSeg#546]
> **Config file**

```
MODEL:
    MODEL_NAME: "hrnet"
    DEFAULT_NORM_TYPE: "bn"
    HRNET:
        STAGE2:
            NUM_CHANNELS: [18, 36]
        STAGE3:
            NUM_CHANNELS: [18, 36, 72]
        STAGE4:
            NUM_CHANNELS: [18, 36, 72, 144]

# ????
TRAIN_CROP_SIZE: (256, 256)
EVAL_CROP_SIZE: (256, 256)
AUG:
    AUG_METHOD: "unpadding"
    FIX_RESIZE_SIZE: (256, 256)
BATCH_SIZE: 200
TRAIN:
    PRETRAINED_MODEL_DIR: "./pretrained_model/hrnet_w18_bn_cityscapes/"
    MODEL_SAVE_DIR: "./saved_model/hrnet_optic/"
    SNAPSHOT_EPOCH: 5
TEST:
    TEST_MODEL: "./saved_model/hrnet_optic/final"
SOLVER:
    NUM_EPOCHS: 50
    LR: 0.01
    LR_POLICY: "poly"
    OPTIMIZER: "adam"
    LOSS: ["lovasz_softmax_loss","softmax_loss"]
    LOSS_WEIGHT:
        LOVASZ_SOFTMAX_LOSS: 0.5
        SOFTMAX_LOSS: 0.5
```
**Error message:**
```
  File "pdseg/train.py", line 474, in <module>
    main(args)
  File "pdseg/train.py", line 457, in main
    train(cfg)
  File "pdseg/train.py", line 239, in train
    train_prog, startup_prog, phase=ModelPhase.TRAIN)
  File "/home/data/yrj/yrj/workplace/pycharm_wp/PaddleSeg/pdseg/models/model_builder.py", line 196, in build_model
    multi_lovasz_softmax_loss(logits, label, mask))
  File "/home/data/yrj/yrj/workplace/pycharm_wp/PaddleSeg/pdseg/lovasz_losses.py", line 226, in multi_lovasz_softmax_loss
    avg_loss = lovasz_softmax(probas, label, ignore=ignore_mask)
  File "/home/data/yrj/yrj/workplace/pycharm_wp/PaddleSeg/pdseg/lovasz_losses.py", line 133, in lovasz_softmax
    loss = lovasz_softmax_flat(vprobas, vlabels, classes=classes)
  File "/home/data/yrj/yrj/workplace/pycharm_wp/PaddleSeg/pdseg/lovasz_losses.py", line 166, in lovasz_softmax_flat
    fg_sorted = fluid.layers.gather(fg, perm)
  File "/home/yrj/softpackage/conda/envs/paddle_env/lib/python3.5/site-packages/paddle/fluid/layers/nn.py", line 8328, in gather
    attrs={'overwrite': overwrite})
  File "/home/yrj/softpackage/conda/envs/paddle_env/lib/python3.5/site-packages/paddle/fluid/layer_helper.py", line 43, in append_op
    return self.main_program.current_block().append_op(*args, **kwargs)
  File "/home/yrj/softpackage/conda/envs/paddle_env/lib/python3.5/site-packages/paddle/fluid/framework.py", line 2917, in append_op
    attrs=kwargs.get("attrs", None))
  File "/home/yrj/softpackage/conda/envs/paddle_env/lib/python3.5/site-packages/paddle/fluid/framework.py", line 2123, in __init__
    self.desc.infer_shape(self.block.desc)
paddle.fluid.core_avx.EnforceNotMet: 

--------------------------------------
C++ Traceback (most recent call last):
--------------------------------------
0   paddle::framework::OpDesc::InferShape(paddle::framework::BlockDesc const&) const
1   paddle::operators::GatherOp::InferShape(paddle::framework::InferShapeContext*) const
2   paddle::platform::EnforceNotMet::EnforceNotMet(std::string const&, char const*, int)
3   paddle::platform::GetCurrentTraceBackString()

----------------------
Error Message Summary:
----------------------
InvalidArgumentError: The last dim of index should be 1 when it is 2D, but we get -1
  [Hint: Expected index_dims[1] == 1, but received index_dims[1]:-1 != 1:1.] (at /paddle/paddle/fluid/operators/gather_op.cc:46)
  [operator < gather > error]
```

> **other**
```
os: ubuntu 18.04
paddle: 2.0rc
paddle_seg: 0.7.0
```

**Comment by lytk01** [comment]
It is ok when I change the code in [lovasz_losses.py](https://github.com/PaddlePaddle/PaddleSeg/blob/9ec92a12b2e467b6fd17eefddad9b70da2a28ed4/pdseg/lovasz_losses.py#L161) from
```
        if C == 1:
            if len(classes_to_sum) > 1:
                raise ValueError('Sigmoid output possible only with 1 class')
            class_pred = probas[:, 0]
        else:
            class_pred = probas[:, c]
        errors = fluid.layers.abs(fg - class_pred)
        errors_sorted, perm = fluid.layers.argsort(
            errors, axis=0, descending=True)
        errors_sorted.stop_gradient = False

        fg_sorted = fluid.layers.gather(fg, perm)
        fg_sorted.stop_gradient = True
```
to 
```
        if C == 1:
            if len(classes_to_sum) > 1:
                raise ValueError('Sigmoid output possible only with 1 class')
            class_pred = probas[:, 0]
        else:
            class_pred = probas[:, c]

        if len(class_pred.shape)<2:
            class_pred=fluid.layers.unsqueeze(class_pred,axes=1)

        errors = fluid.layers.abs(fg - class_pred)
        errors_sorted, perm = fluid.layers.argsort(
            errors, axis=0, descending=True)
        errors_sorted.stop_gradient = False

        fg_sorted = fluid.layers.gather(fg, perm)
        fg_sorted.stop_gradient = True
```
,but I don't know it's correct.

**Comment by choimroc** [comment]
@lytk01 I had the same problem , have your changes been verified?

**Comment by LutaoChu** [comment]
你好，PaddleSeg已推出动态图版本Lovasz loss，欢迎体验！
https://github.com/PaddlePaddle/PaddleSeg/blob/develop/docs/lovasz_loss.md

**Comment by LutaoChu** [comment]
> It is ok when I change the code in [lovasz_losses.py](https://github.com/PaddlePaddle/PaddleSeg/blob/9ec92a12b2e467b6fd17eefddad9b70da2a28ed4/pdseg/lovasz_losses.py#L161) from
> 
> ```
>         if C == 1:
>             if len(classes_to_sum) > 1:
>                 raise ValueError('Sigmoid output possible only with 1 class')
>             class_pred = probas[:, 0]
>         else:
>             class_pred = probas[:, c]
>         errors = fluid.layers.abs(fg - class_pred)
>         errors_sorted, perm = fluid.layers.argsort(
>             errors, axis=0, descending=True)
>         errors_sorted.stop_gradient = False
> 
>         fg_sorted = fluid.layers.gather(fg, perm)
>         fg_sorted.stop_gradient = True
> ```
> 
> to
> 
> ```
>         if C == 1:
>             if len(classes_to_sum) > 1:
>                 raise ValueError('Sigmoid output possible only with 1 class')
>             class_pred = probas[:, 0]
>         else:
>             class_pred = probas[:, c]
> 
>         if len(class_pred.shape)<2:
>             class_pred=fluid.layers.unsqueeze(class_pred,axes=1)
> 
>         errors = fluid.layers.abs(fg - class_pred)
>         errors_sorted, perm = fluid.layers.argsort(
>             errors, axis=0, descending=True)
>         errors_sorted.stop_gradient = False
> 
>         fg_sorted = fluid.layers.gather(fg, perm)
>         fg_sorted.stop_gradient = True
> ```
> 
> ,but I don't know it's correct.

It's correct. We have fixed it in #849
